README.md

```diff
@@ -0,0 +1,71 @@
+[![Build Status](https://github.com/gafter/AutoHashEqualsCached.jl/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/gafter/AutoHashEqualsCached.jl/actions/workflows/CI.yml?query=branch%3Amain)
+
+# AutoHashEqualsCached
+
+A pair of macros to add `==` and `hash()` to struct types: `@auto_hash_equals` and `@auto_hash_equals_cached`.
+
+# `@auto_hash_equals`
+
+The macro `@auto_hash_equals` produces an implementation of `Base.hash(x)` that computes the hash code when invoked.  It is offered for compatibility with the package [AutoHashEquals](https://github.com/andrewcooke/AutoHashEquals.jl).
+
+You use it like so:
+
+```julia
+@auto_hash_equals struct Box
+    x
+end
+```
+
+which is translated to
+
+```julia
+struct Box
+    x
+end
+Base.hash(x::Type, h::UInt) = hash(x.x, hash(:Foo, h))
+Base.(:(==))(a::Box, b::Box) = isequal(a.x, b.x)
+```
+
+However, there are a few enhancements beyond what is provided by [AutoHashEquals](https://github.com/andrewcooke/AutoHashEquals.jl):
+
+- It handles empty structs without error
+- You can use it either before or after another macro.
+
+For compatibility with [AutoHashEquals](https://github.com/andrewcooke/AutoHashEquals.jl), we do not take the type arguments of a generic type into account in `==`.  So a `Box{Int}(1)` will test equal to a `Box{Any}(1)`.
+
+# `@auto_hash_equals_cached`
+
+The macro `@auto_hash_equals_cached` is useful for non-mutable struct types that define recursive or deep data structures (and therefore are likely to be stored on the heap).  It computes the hash code during construction and caches it in a field of the struct.  If you are working with data structures of any significant depth, computing the hash once can speed things up at the expense of one additional field per struct.
+
+You use it like so:
+
+```julia
+@auto_hash_equals_cached struct Point{T<:Any, U<:Any}
+  x::T
+  y::U
+end
+```
+
+which is translated to
+
+```julia
+struct Point{T<:Any, U<:Any}
+    x::T
+    y::U
+    _cached_hash::UInt
+    Point{T,U}(x::T, y::U) where {T<:Any, U<:Any} = new{T,U}(x,y,hash(y, hash(x, hash(Point{T,U}))))
+end
```

Comment:
Julia should pad so that a `UInt` is properly aligned.  I believe Julia uses same struct layout as C does.  Since one of the fields is a `UInt`, the struct as a whole must be padded to a multiple of that size.  Adding this field forces that padding to be before the newly added field.  For example, if the contents of the struct used to be `x::Int8`, the whole struct would have only taken up a single byte and had byte alignment.  But the new field is 64-bits (8 bytes), so there would be 7 bytes of padding before this field, making the total struct size 16 bytes and alignment 8 bytes.  That is the worse case.  Placing the field at the beginning wouldn't make it any better.

One advantage of putting it at the end is that it was easier to make it interoperate with `@match`.  `@mutts_interface` places its field at the beginning.

Reply:
Cool, thanks. That exactly matches my analysis as well 👍